spec/mongoid/criteria_spec.rb

```diff
@@ -3422,20 +3422,111 @@ def self.ages; self; end
 
       context "when querying on a big decimal" do
 
-        let(:sales) do
-          BigDecimal('0.1')
+        context 'when map_big_decimal_to_decimal128 is false' do
+
+          around do |example|
+            current = Mongoid.map_big_decimal_to_decimal128
+            Mongoid.map_big_decimal_to_decimal128 = false
+            example.run
+            Mongoid.map_big_decimal_to_decimal128 = current
+          end
+
+          let(:sales) do
+            BigDecimal('0.1')
+          end
+
+          let!(:band) do
+            Band.create!(name: "Boards of Canada", sales: sales)
+          end
+
+          let(:from_db) do
+            Band.where(sales: sales).first
+          end
+
+          it "finds the document by the big decimal value" do
+            expect(from_db).to eq(band)
+          end
         end
 
-        let!(:band) do
-          Band.create!(name: "Boards of Canada", sales: sales)
+        context 'when map_big_decimal_to_decimal128 is true' do
+
+          around do |example|
+            current = Mongoid.map_big_decimal_to_decimal128
+            Mongoid.map_big_decimal_to_decimal128 = true
+            example.run
+            Mongoid.map_big_decimal_to_decimal128 = current
+          end
+
+          let(:sales) do
+            BigDecimal('0.1')
+          end
+
+          let!(:band) do
+            Band.create!(name: "Boards of Canada", sales: sales)
+          end
+
+          let(:from_db) do
+            Band.where(sales: sales).first
+          end
+
+          it "finds the document by the big decimal value" do
+            expect(from_db).to eq(band)
+          end
         end
+      end
 
-        let(:from_db) do
-          Band.where(sales: sales).first
+      context "when querying on a big decimal from a dynamic field" do
+
+        context 'when map_big_decimal_to_decimal128 is false' do
+
+          around do |example|
+            current = Mongoid.map_big_decimal_to_decimal128
+            Mongoid.map_big_decimal_to_decimal128 = false
+            example.run
+            Mongoid.map_big_decimal_to_decimal128 = current
+          end
+
+          let(:fans) do
+            BigDecimal('139432.0002')
+          end
+
+          let!(:band) do
+            Band.create!(name: "Boards of Canada", fans: fans)
+          end
+
+          let(:from_db) do
+            Band.where(fans: fans.to_s).first
```

Comment:
This is an interesting case. A `BigDecimal` was inserted with `map_big_decimal_to_decimal128=false` so it was stored as a string. Now, if we try to query it using a `BigDecimal`, it won't find it. Note this field is _dynamic_ so it can't be expected to know that strings are also `BigDecimal`s. In the case where we set the type of the field to `BigDecimal`, it does find the document, in this case I don't think it should...

What do you think @p-mongo?

Reply:
This caveat can go into the BigDecimal field documentation. If users query on the field, they are would need to either query with both value types (BigDecimal and string, using e.g. `$in`) or they would need to migrate their data to Decimal128 storage.

Reply:
done